web/source/osk/visualKeyboard.ts

```diff
@@ -161,9 +161,23 @@ namespace com.keyman.osk {
     protected renameSpecialKey(oldText: string): string {
       let keyman = (<KeymanBase>window['keyman'])
       // If a 'special key' mapping exists for the text, replace it with its corresponding special OSK character.
-      return VisualKeyboard.specialCharacters[oldText] ?
-        String.fromCharCode(0XE000 + VisualKeyboard.specialCharacters[oldText]) :
-        oldText;
+      if(oldText != "*ZWNJ*") {
+        return VisualKeyboard.specialCharacters[oldText] ?
+          String.fromCharCode(0XE000 + VisualKeyboard.specialCharacters[oldText]) :
+          oldText;
+      } else {
+        // Examine platform; choose appropriate symbol.
+        let coreSpec = keyman.util.device.coreSpec;
+        let zeroWidthPUA = 0XE000 + VisualKeyboard.specialCharacters[oldText];
+
+        // Default ZWNJ symbol comes from iOS.  We'd rather match the system defaults where
+        // possible / available though, and there's a different standard symbol on Android.
+        if(coreSpec.OS == com.keyman.utils.OperatingSystem.Android) {
+          zeroWidthPUA += 1; // Use the Android version.
+        } 
+
+        return String.fromCharCode(zeroWidthPUA);
+      }
```

Comment:
I think I'd prefer the following pattern (and corresponding update to `specialCharacters` static):

1. It's shorter 😉
2. We explicitly define the map for the two symbols, which is easier to maintain
3. It allows for a keyboard dev to override the 'magic' `*ZWNJ*` with the specific symbol if they really want to.

```suggestion
      if(oldText == "*ZWNJ*") {
        // Default ZWNJ symbol comes from iOS.  We'd rather match the system defaults where
        // possible / available though, and there's a different standard symbol on Android.
        oldText = keyman.util.device.coreSpec.OS == com.keyman.utils.OperatingSystem.Android ?
          "*ZWNJAndroid*" :
          "*ZWNJiOS*";
      }
     
      let zeroWidthPUA = 0XE000 + VisualKeyboard.specialCharacters[oldText];
        
      return VisualKeyboard.specialCharacters[oldText] ?
        String.fromCharCode(0XE000 + VisualKeyboard.specialCharacters[oldText]) :
        oldText;
```

Reply:
```keyman.util.device.coreSpec.OS == com.keyman.utils.OperatingSystem.Android``` makes me squirm a little. Is it really right that we have `keyman.util.device` but `com.keyman.utils.OperatingSystem`? Is this legacy coming out to play again?

Reply:
The latter is Web-core compatible, while the former is not.  The old `kmwdevice.ts` had too many DOM-dependent ties, so it was kept around as 'legacy', and for device-detection.  Could check against the raw string that `keyman.util.device` uses instead of the Web-core enum, though.

Reply:
> 3. It allows for a keyboard dev to override the 'magic' *ZWNJ* with the specific symbol if they really want to.

It does?  I'm not seeing the part where a developer can specify a general `*ZWNJ*` before your suggested code forces it to either `*ZWNJAndroid*` or `*ZWNJiOS*`.

Reply:
They can use `*ZWNJAndroid*` or `*ZWNJiOS*` in their keyboard.

Reply:
> Could check against the raw string that `keyman.util.device` uses instead of the Web-core enum, though.

Let's stick with using the Web-core enum!